runtime/common/src/parachains.rs

```diff
@@ -142,6 +231,145 @@ pub enum Origin {
 	Parachain(ParaId),
 }
 
+/// An offence that is filed if the validator has submitted a double vote.
+#[derive(RuntimeDebug)]
+#[cfg_attr(feature = "std", derive(Clone, PartialEq, Eq))]
+pub struct DoubleVoteOffence<Offender> {
+	/// The current session index in which we report a validator.
+	session_index: SessionIndex,
+	/// The size of the validator set in current session/era.
+	validator_set_count: u32,
+	/// An offender that has submitted two conflicting votes.
+	offender: Offender,
+}
+
+impl<Offender: Clone> Offence<Offender> for DoubleVoteOffence<Offender> {
+	const ID: Kind = *b"double-vote:doub";
+	type TimeSlot = SessionIndex;
+
+	fn offenders(&self) -> Vec<Offender> {
+		vec![self.offender.clone()]
+	}
+
+	fn session_index(&self) -> SessionIndex {
+		self.session_index
+	}
+
+	fn validator_set_count(&self) -> u32 {
+		self.validator_set_count
+	}
+
+	fn time_slot(&self) -> Self::TimeSlot {
+		self.session_index
+	}
+
+	fn slash_fraction(_offenders_count: u32, _validator_set_count: u32) -> Perbill {
+		// Slash 100%.
+		Perbill::from_percent(100)
+	}
+}
+
+impl<O: Clone> ParachainsOffence<O> for DoubleVoteOffence<O> {
+	fn new(
+		session_index: SessionIndex,
+		validator_set_count: u32,
+		offender: O,
+	) -> Self {
+		Self {
+			session_index,
+			validator_set_count,
+			offender,
+		}
+	}
+}
+
+pub trait ParachainsOffence<O>: Offence<O> {
```

Comment:
Missing docs. I don't understand what this trait is for, but a trait with nothing but a constructor is a bit strange.